proposals/0064-transaction-receipt.md

```diff
@@ -0,0 +1,416 @@
+
+- Definitions:
+  - The `intermediate_root` is the 32 byte root of the binary merkle tree
+    as externally specified.
+    [Specification: Binary Merkle Tree](https://github.com/solana-foundation/specs/blob/main/core/merkle-tree.md)
+  - The `root_prefix` is the byte `0x80`
+  - The function `u64_le_encode` encodes a 64-bit integer in little-endian
+    byte order.
+  - The `leaf_count` is the number of TransactionReceiptData objects to
+    serve as leaf nodes in the tree.
+
+- If the leaf count is zero, the output is
+  `sha256(root_prefix || u64_le_encode(0))`.
+
+- If the leaf count is non-zero, the output is
+  `sha256(root_prefix || intermediate_root || u64_le_encode(leaf_count))`.
+
+#### Tree design considerations
+
+- *Use of the PoH hash tree construction*
+  - Allows reusing existing code for constructing the transaction receipt tree.
+  - Avoids introducing a new cryptographic construction.
+  - Alternatives to SHA-256 based constructions, such as BLAKE3 or SHA-3 would
+    offer superior theoretical performance.
+  - Existing optimized SHA-256 hash tree implementations are readily available.
+  - As of 2023-Oct, hardware implementations of SHA-256 are currently available
+    (via SHA-NI on x86, via Firedancer's SystemVerilog implementation on AWS F1
+    FPGA), whereas BLAKE3 and SHA-3 are not. (SHA-3 is only available on recent
+    Arm cores)
+
+- *Leaf count suffix*:
+  The PoH tree implicitly expands the internal leaf count to a power of two,
+  causing the `intermediate_root` two have more than one pre-image for certain
+  leaf counts.  This is fixed by the leaf count suffix in the final hash.
+
+- *Exclusion proofs are not provided*:
+  Although it is possible to construct Merkle-based set exclusion proofs, this
+  feature was not part of this proposal's design criteria.  A proposed exclusion
+  proof mechanism involves sorting transaction receipts by their message hash.
+  This change may be considered in a future version of the transaction receipt
+  tree.
+
+#### Examples
+
+The following illustrates transaction receipt hash trees for various inputs.
+
+```txt
+Transaction receipt tree with an empty vector of transaction receipts
+where Nα is the root.
+Nα := sha256(0x80 || u64_le_encode(0))
+
+Nα
+|
+0
+```
+
+```txt
+Transaction receipt tree with four transaction receipts as leaf
+nodes [L0, L1, L2, L3] where R0, R1, R2, R3 are the transaction 
+receipts and Nδ is the root.
+           Nδ
+          /  \
+        /     \
+       Nγ      N(transaction_receipts)
+      /  \
+     /    \
+   Nα      Nβ
+  / |     / \
+ /  |    /   \
+L0  L1   L2  L3
+
+L0 := sha256(concat(0x00, R0))
+L1 := sha256(concat(0x00, R1))
+L2 := sha256(concat(0x00, R2))
+L3 := sha256(concat(0x00, R3))
+Nα := sha256(concat(0x01, hash(L0), hash(L1)))
+Nβ := sha256(concat(0x01, hash(L2), hash(L3)))
+Nγ := sha256(concat(0x01, hash(Nα), hash(Nβ)))
+Nδ := sha256(concat(0x80, hash(Nγ),len([L0, L1, L2, L3])))
```

Comment:
> why do we need to rehash the previous layers' nodes? or alternatively, why _don't_ we need to double-hash the final root?

@t-nelson can you elaborate here? As Sam-Kim suggested we hash the intermediate root with the length to prevent length extension attacks. I don't see a reason to double-hash the final root.

Reply:
why is it
```
Nα := sha256(concat(0x01, hash(L0), hash(L1)))
```
and not
```
Nα := sha256(concat(0x01, L0, L1))
```
?

Reply:
> why is it
> 
> ```
> Nα := sha256(concat(0x01, hash(L0), hash(L1)))
> ```
> 
> and not
> 
> ```
> Nα := sha256(concat(0x01, L0, L1))
> ```
> 
> ?

Right, this should be the latter not the former. Let me correct it

Reply:
@t-nelson does that mean the solana merkle tree spec here is wrong? https://github.com/solana-foundation/specs/blob/main/core/merkle-tree.md#canonical-construction
<img width="755" alt="Screenshot 2023-10-19 at 12 09 52 PM" src="https://github.com/solana-foundation/solana-improvement-documents/assets/32778608/de1667a2-0769-4cbb-8db3-cdf29b38013e">

Reply:
Confirmed this @ripatel-fd, it wouldn't make sense to do what you're suggesting as it would not remain a hash tree. The intermediate node is always a hash of the vector of child/leaf nodes.